app/components/form/FormNavGuard.tsx

```diff
@@ -0,0 +1,63 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { useEffect } from 'react'
+import { type FieldValues, type UseFormReturn } from 'react-hook-form'
+import { useBlocker } from 'react-router-dom'
+
+import { Modal } from '~/ui/lib/Modal'
+
+export function FormNavGuard<TFieldValues extends FieldValues>({
+  form,
+}: {
+  form: UseFormReturn<TFieldValues>
```

Comment:
This doesn't need to be generic since we're not actually using the form values. I think you can do `form: UseFormReturn` with no param and it'll work.

Reply:
Both of the callsites for this have `form` as a `UseFormReturn<TFieldValues>`. I think it'd be easier to just pass it in as-is and set the type within the new component to also have the generic? (TS complains at the callsites if I strip it from the new component's type) Or is there an aspect to generics that I'm missing?

Reply:
Yep, I was wrong. I thought that leaving it off like this 

```diff
-export function FormNavGuard<TFieldValues extends FieldValues>({
-  form,
-}: {
-  form: UseFormReturn<TFieldValues>
-}) {
+export function FormNavGuard({ form }: { form: UseFormReturn }) {
   const { isDirty, isSubmitting, isSubmitSuccessful } = form.formState
   // Confirms with the user if they want to navigate away if the form is
   // dirty. Does not intercept everything e.g. refreshes or closing the tab
```

would be more like `UseFormReturn<any>` (which does work btw because it doesn't care about types anymore after), but it's actually `UseFormReturn<FieldValues>` (the default value for that generic param) and surprisingly, that is incompatible with our `TFieldValues extends FieldValues`. I think you're right, the best thing is to leave it as-is.